src/core/lib/channel/metrics.h

```diff
@@ -61,67 +65,78 @@ class GlobalInstrumentsRegistry {
     // runs or between different versions.
     uint32_t index;
   };
-  struct GlobalUInt64CounterHandle : GlobalHandle {};
-  struct GlobalDoubleCounterHandle : GlobalHandle {};
-  struct GlobalUInt64HistogramHandle : GlobalHandle {};
-  struct GlobalDoubleHistogramHandle : GlobalHandle {};
+  struct GlobalUInt64CounterHandle : public GlobalHandle {};
+  struct GlobalDoubleCounterHandle : public GlobalHandle {};
+  struct GlobalUInt64HistogramHandle : public GlobalHandle {};
+  struct GlobalDoubleHistogramHandle : public GlobalHandle {};
 
   // Creates instrument in the GlobalInstrumentsRegistry.
   static GlobalUInt64CounterHandle RegisterUInt64Counter(
       absl::string_view name, absl::string_view description,
-      absl::string_view unit, std::vector<absl::string_view> label_keys,
-      std::vector<absl::string_view> optional_label_keys);
+      absl::string_view unit, absl::Span<const absl::string_view> label_keys,
+      absl::Span<const absl::string_view> optional_label_keys);
   static GlobalDoubleCounterHandle RegisterDoubleCounter(
       absl::string_view name, absl::string_view description,
-      absl::string_view unit, std::vector<absl::string_view> label_keys,
-      std::vector<absl::string_view> optional_label_keys);
+      absl::string_view unit, absl::Span<const absl::string_view> label_keys,
+      absl::Span<const absl::string_view> optional_label_keys);
   static GlobalUInt64HistogramHandle RegisterUInt64Histogram(
       absl::string_view name, absl::string_view description,
-      absl::string_view unit, std::vector<absl::string_view> label_keys,
-      std::vector<absl::string_view> optional_label_keys);
+      absl::string_view unit, absl::Span<const absl::string_view> label_keys,
+      absl::Span<const absl::string_view> optional_label_keys);
   static GlobalDoubleHistogramHandle RegisterDoubleHistogram(
       absl::string_view name, absl::string_view description,
-      absl::string_view unit, std::vector<absl::string_view> label_keys,
-      std::vector<absl::string_view> optional_label_keys);
-  // Getter functions for StatsPlugins to query registered counters/histograms.
-  static const std::vector<GlobalInstrumentDescriptor>& instruments();
+      absl::string_view unit, absl::Span<const absl::string_view> label_keys,
+      absl::Span<const absl::string_view> optional_label_keys);
+  static void ForEach(
+      absl::AnyInvocable<void(const GlobalInstrumentDescriptor&)> f);
 
   GlobalInstrumentsRegistry() = delete;
 
  private:
-  static std::vector<GlobalInstrumentDescriptor> instruments_;
+  // Uses the Construct-on-First-Use idiom to avoid the static initialization
+  // order fiasco.
+  static std::vector<GlobalInstrumentDescriptor>& gInstruments() {
```

Comment:
I think we need to use this pattern to avoid the Static Initialization Order Fiasco problem: https://en.cppreference.com/w/cpp/language/siof. IIUC, having a static class member such as `static std::vector<GlobalInstrumentDescriptor> instruments_;` means it will be default initialized (calling its default constructor). And IIUC default initialization is not part of static initialization: https://en.cppreference.com/w/cpp/language/initialization#Static_initialization. So it is part of dynamic initialization and the order is indeterminately sequenced across different translation units. Thus we need to use pattern.

Reply:
Oh, interesting -- so I guess the reason we need this here but don't need it for the stats plugin registry is that we don't expect stats plugins to be registered before static initialization is complete.  If so, that sounds fine.

Reply:
Yeah, IIUC we expect StatsPlugins to be registered after the main function begins.